### PR TITLE
use multiscalar_mul for CLSAG

### PR DESCRIPTION
**Pull request by jbird186**
Approximate 5 - 10% speedup for signing, and 25 - 30% speedup for verification.

### CODE REVIEW

File: coins/monero/src/ringct/clsag/mod.rs [hunk 1]
```diff
@@ -100,6 +100,7 @@ fn core(
 ) -> ((EdwardsPoint, Scalar, Scalar), Scalar) {
   let n = ring.len();
 
+  let base_precomp = VartimeEdwardsPrecomputation::new([ED25519_BASEPOINT_POINT]);
```
Comment (kayabaNerve) [inline coins/monero/src/ringct/clsag/mod.rs]: I can't dislike this PR if it brings performance up, yet I have to ask if this line actually improves performance. This is recalculating ED25519_BASEPOINT_TABLE, presumably at a smaller size than that table.

I do get it offers the below API, which the below table doesn't. I'd legitimately suggest benchmarking it as `(s[i] * TABLE) + ...` though. The addition is incredibly cheap, so the argument has to be this API (increasing memory, costing doubles in setup, removing point doubles inside the loop) is still better than the no-setup presumably-less-doubles-needed  version based on using the existing table.

That isn't to mention how this is calculated unconditionally yet only used with Mode::Verify (meaning it likely should be within an Option).

File: coins/monero/src/ringct/clsag/mod.rs [hunk 2]
```diff
@@ -174,10 +175,24 @@ fn core(
...
+    };
+
     let PH = hash_to_point(&P[i]);
-    // Shouldn't be an issue as all of the variables in this vartime statement are public
-    let R = (s[i] * PH) + images_precomp.vartime_multiscalar_mul([c_p, c_c]);
+
+    // (c_p * I) + (c_c * D) + (s_i * PH)
+    let R = match A_c1 {
+      // Shouldn't be an issue as all of the variables in the sign vartime statement are public
+      Mode::Sign(..) => images_precomp.vartime_multiscalar_mul([c_p, c_c]) + (s[i] * PH),
```
Comment (kayabaNerve) [inline coins/monero/src/ringct/clsag/mod.rs]: If we are now diversifying this, I'd probably suggest we should use a constant time call here and be safe than have to justify this

File: coins/monero/src/ringct/clsag/mod.rs (hunk 1 above)
Comment (kayabaNerve) [reply coins/monero/src/ringct/clsag/mod.rs]: Apologies if this is effectively asking we add a criterion benchmark and multiple different variants for it :sweat_smile:  Again, I can fully acknowledge the performance is better, even if it's not fully optimal, and discuss moving forward on that premise alone.

File: coins/monero/src/ringct/clsag/mod.rs (hunk 2 above)
Comment (jbird186) [reply coins/monero/src/ringct/clsag/mod.rs]: Done.

File: coins/monero/src/ringct/clsag/mod.rs (hunk 1 above)
Comment (jbird186) [reply coins/monero/src/ringct/clsag/mod.rs]: I wrote some spaghetti code (see below) for quick benchmarking on calculating `L`. I think this goes without saying, but obviously these results might not perfectly translate into real CLSAG usage.

Results (ms):

|   | 1 Ring | 25 Rings |
| ------------- | ------------- | ------------- |
| Original | 0.992 | 24.626 |
| Full vartime_multiscalar_mul | 0.483 | 12.116 |
| Basepoint Table | 0.611 | 15.284 |
| Precomp | 0.471 | 11.874 |
| Precomp w/ OnceLock | 0.474 | 11.613 |

Using `OnceLock` (in the style of `INV_EIGHT()`) was something that I originally didn't consider. For one ring it's essentially a wash, but for several rings (where performance really matters, anyway), there's a slight improvement. Going off of these results, "Precomp w/ OnceLock" is probably the best option.

I'll leave it up to you as to which method should be used, or if more benchmarking should be done. I've changed the PR into a draft for now.

Benchmarks:

```
#![allow(non_snake_case)]

use std_shims::sync::OnceLock;
use criterion::{black_box, criterion_main, Criterion};
use curve25519_dalek::{
  constants::{ED25519_BASEPOINT_TABLE, ED25519_BASEPOINT_POINT},
  traits::{VartimeMultiscalarMul, VartimePrecomputedMultiscalarMul},
  edwards::{EdwardsPoint, VartimeEdwardsPrecomputation},
};
use monero_serai::{
  hash_to_scalar, ringct::raw_hash_to_point,
};

const RING_SIZE: usize = 16;

fn basepoint() {
    let mut c = Criterion::default();

    let c_p = hash_to_scalar(b"test data 1");
    let c_c = hash_to_scalar(b"test data 2");
    let s_i = hash_to_scalar(b"test data 3");
    let P_i = raw_hash_to_point([10; 32]);
    let C_i = raw_hash_to_point([11; 32]);

    c.bench_function("original (1 ring)", |b| {
        b.iter(||
            for _ in 0..RING_SIZE {
                black_box((&s_i * ED25519_BASEPOINT_TABLE) + (c_p * P_i) + (c_c * C_i));
            }
        )
    });

    c.bench_function("original (25 rings)", |b| {
        b.iter(||
            for _ in 0..25 * RING_SIZE {
                black_box((&s_i * ED25519_BASEPOINT_TABLE) + (c_p * P_i) + (c_c * C_i));
            }
        )
    });

    c.bench_function("full vartime_multiscalar_mul (1 ring)", |b| {
        b.iter(|| {
            for _ in 0..RING_SIZE {
                black_box(EdwardsPoint::vartime_multiscalar_mul([s_i, c_p, c_c], [ED25519_BASEPOINT_POINT, P_i, C_i]));
            }
        })
    });

    c.bench_function("full vartime_multiscalar_mul (25 rings)", |b| {
        b.iter(|| {
            for _ in 0..25 * RING_SIZE {
                black_box(EdwardsPoint::vartime_multiscalar_mul([s_i, c_p, c_c], [ED25519_BASEPOINT_POINT, P_i, C_i]));
            }
        })
    });

    c.bench_function("basepoint table (1 ring)", |b| {
        b.iter(|| {
            for _ in 0..RING_SIZE {
                black_box((&s_i * ED25519_BASEPOINT_TABLE) + EdwardsPoint::vartime_multiscalar_mul([c_p, c_c], [P_i, C_i]));
            }
        })
    });

    c.bench_function("basepoint table (25 rings)", |b| {
        b.iter(|| {
            for _ in 0..25 * RING_SIZE {
                black_box((&s_i * ED25519_BASEPOINT_TABLE) + EdwardsPoint::vartime_multiscalar_mul([c_p, c_c], [P_i, C_i]));
            }
        })
    });

    c.bench_function("precomp (1 ring)", |b| {
        b.iter(|| {
            let base_precomp = VartimeEdwardsPrecomputation::new([ED25519_BASEPOINT_POINT]);
            for _ in 0..RING_SIZE {
                black_box(base_precomp.vartime_mixed_multiscalar_mul([s_i], [c_p, c_c], [P_i, C_i]));
            }
        })
    });

    c.bench_function("precomp (25 rings)", |b| {
        b.iter(|| {
            for _ in 0..25 {
                let base_precomp = VartimeEdwardsPrecomputation::new([ED25519_BASEPOINT_POINT]);
                for _ in 0..RING_SIZE {
                    black_box(base_precomp.vartime_mixed_multiscalar_mul([s_i], [c_p, c_c], [P_i, C_i]));
                }
            }
        })
    });

    c.bench_function("precomp w/ OnceLock (1 ring)", |b| {
        b.iter(|| {
            static PRECOMP_CELL: OnceLock<VartimeEdwardsPrecomputation> = OnceLock::new();
            fn PRECOMP() -> &'static VartimeEdwardsPrecomputation {
                PRECOMP_CELL.get_or_init(|| VartimeEdwardsPrecomputation::new([ED25519_BASEPOINT_POINT]))
            }
            for _ in 0..RING_SIZE {
                black_box(PRECOMP().vartime_mixed_multiscalar_mul([s_i], [c_p, c_c], [P_i, C_i]));
            }
        })
    });

    c.bench_function("precomp w/ OnceLock (25 rings)", |b| {
        b.iter(|| {
            static PRECOMP_CELL: OnceLock<VartimeEdwardsPrecomputation> = OnceLock::new();
            fn PRECOMP() -> &'static VartimeEdwardsPrecomputation {
                PRECOMP_CELL.get_or_init(|| VartimeEdwardsPrecomputation::new([ED25519_BASEPOINT_POINT]))
            }
            for _ in 0..25 * RING_SIZE {
                black_box(PRECOMP().vartime_mixed_multiscalar_mul([s_i], [c_p, c_c], [P_i, C_i]));
            }
        })
    });
}
criterion_main!(basepoint);
```

Comment (kayabaNerve) [reply coins/monero/src/ringct/clsag/mod.rs]: Using a OnceLock sounds good then :+1:

Thanks for being thorough on this and incorporating my feedback :) I definitely don't want to turn away better code for not being optimal, Seems like overall, this'll end up +4% off original without a penalty on sign? ^_^

Comment (jbird186) [reply coins/monero/src/ringct/clsag/mod.rs]: Thanks! Should be good to go now.

> Seems like overall, this'll end up +4% off original without a penalty on sign?

For just this change? If so, yeah, it's a few percentage. This PR as a whole though should be much higher for that, for both signing and verifying.

**Comment by kayabaNerve** [comment]
LGTM! :) Will merge when CI passes.

File: coins/monero/src/ringct/clsag/mod.rs (hunk 1 above)
Comment (kayabaNerve) [reply coins/monero/src/ringct/clsag/mod.rs]: Yep! Sorry for the confusion there.

**Pull request merged**